src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Windows.cs

```diff
@@ -17,6 +17,85 @@ public partial class Socket
 
         internal void ReplaceHandleIfNecessaryAfterFailedConnect() { /* nop on Windows */ }
 
+        public Socket(SocketInformation socketInformation)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            InitializeSockets();
+
+            SocketError errorCode = SocketPal.CreateSocket(socketInformation, out _handle,
+                ref _addressFamily, ref _socketType, ref _protocolType);
+
+            if (errorCode != SocketError.Success)
+            {
+                Debug.Assert(_handle.IsInvalid);
+
+                if (errorCode == SocketError.InvalidArgument)
+                {
+                    throw new ArgumentException(SR.net_sockets_invalid_socketinformation, nameof(socketInformation));
+                }
+
+                // Failed to create the socket, throw.
+                throw new SocketException((int)errorCode);
+            }
+
+            if (_handle.IsInvalid)
+            {
+                throw new SocketException();
+            }
+
+            if (_addressFamily != AddressFamily.InterNetwork && _addressFamily != AddressFamily.InterNetworkV6)
+            {
+                throw new NotSupportedException(SR.net_invalidversion);
+            }
+
+            _isConnected = socketInformation.GetOption(SocketInformationOptions.Connected);
+            _willBlock = !socketInformation.GetOption(SocketInformationOptions.NonBlocking);
+            InternalSetBlocking(_willBlock);
+            _isListening = socketInformation.GetOption(SocketInformationOptions.Listening);
+
+            IPAddress tempAddress = _addressFamily == AddressFamily.InterNetwork ? IPAddress.Any : IPAddress.IPv6Any;
+            IPEndPoint ep = new IPEndPoint(tempAddress, 0);
+
+            Internals.SocketAddress socketAddress = IPEndPointExtensions.Serialize(ep);
+            errorCode = SocketPal.GetSockName(_handle, socketAddress.Buffer, ref socketAddress.InternalSize);
+            if (errorCode == SocketError.Success)
+            {
+                try
+                {
+                    _rightEndPoint = ep.Create(socketAddress);
+                }
+                catch
+                {
```

Comment:
It looks suspicious that we silently suppress all exceptions here. What is the motivation?

Reply:
I seems to be a bad pattern inherited from the old framework code. Would make sense to replace `IPEndPointExtensions.Create()` with a non-throwing `IPEndPointExtensions.TryCreate()`, but I did not want to apply changes of that extent in this PR.

Reply:
If this fails, the socket will be in a bad state. I'd rather we just let it throw.

That said, I don't know why this would fail as I'd expect socket duplication to not work if the address is wrong.

Reply:
Doesn't that apply for all call sites? If we change the pattern, it's better to alter usages everywhere.
I would prefer to change it in a separate clean-up PR everywhere rather than breaking the (bad) consistency on a single call site.

Reply:
Can you point to what other code you're comparing to? UDP would function mostly fine like this but to my knowledge this would be the first API where we can have a connected socket without an endpoint set, and the other common connection-oriented APIs (send/recv) don't not set it so it would never get set.

Failing fast here would be better than failing down the line with a null endpoint, especially if it's "free" because the exception is already getting thrown.

Reply:
The [same exception swallowing pattern](https://github.com/dotnet/runtime/blob/5b617e604cca42aac512f6be809d2a5330586400/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L3141) is present in Send/Receive implementations. I don't get the purpose of those calls though, since we are not doing anything with the `remoteEp` instances being returned.

Failing fast would be indeed better here, however the handling of (undocumented & impossiblish) exceptions coming from internals looks quite ad-hoc in existing code. Some call sites just let them leak out, others prefer to swallow them. Shouldn't be there a guideline point about this?

Reply:
Removed the try-catch, but I'd really like to have better understanding. The code was taken from Framework.